E2e test for cleaning of terminated containers

KevinTMtz opened the pull request:
Added a new e2e test to verify that the node responds to disk pressure by cleaning up containers from terminated pods without evicting running pods, when reclaiming their resources freed enough space to eliminate the disk pressure

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Prevent running pods to be evicted when having disk pressure on the node, if it is possible to eliminate resource pressure evicting terminated pods. 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #128790

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

KevinTMtz committed:
Evict terminated pods on disk pressure

Before evicting running pods, terminal phase pods will be evicted first

KevinTMtz commented:
/cc @AnishShah 
/cc @yujuhong

KevinTMtz committed:
Test terminated pods are evicted on disk pressure

kannon92 commented:
/ok-to-test

cici37 commented:
/remove-sig api-machinery

KevinTMtz committed:
TerminatedPodsEvictionOnDiskPressure e2e node test

KevinTMtz commented:
/test pull-kubernetes-e2e-gce

AnishShah commented on test/e2e_node/eviction_test.go:
```diff
@@ -103,6 +104,53 @@ var _ = SIGDescribe("InodeEviction", framework.WithSlow(), framework.WithSerial(
+			// This is the terminated pod, runEvictionTest will wait for it to finish
+			{
+				evictionPriority: 1,
+				// This is probably working because empty dir is being cleared when the
```
This probably means that it is not testing what we want?

Maybe we should replicate what we do in `ImageGCNoEviction` test below.

Create 2 pods:
1. running pod with 110000 files
2. terminated pod with 110000 files in rootfs (not emptyDir)

only the second pod should be evicted due to exceeding inodes threshold.

KevinTMtz replied:
I added a new test with the following config:
- running pod with 80,000 inods
- terminated pod with 30,000 inods in rootfs (not emptyDir)

This causes the eviction threshold (100,000 inodes) to be passed with a total of 110,000 inodes, making the node to react evicting all terminated pods, freeing the 30,000 inodes written by the terminated pod and eliminating the disk pressure without affecting the running pod

KevinTMtz committed:
Add ImageGCTerminatedPodsEviction e2e node test

Removed TerminatedPodsEvictionOnDiskPressure, since it did not test the expected functionality

gofmt ./pkg/kubelet/eviction/eviction_manager_test.go

KevinTMtz committed:
Remove terminated pods eviction code

KevinTMtz commented:
Since the terminated containers were already being cleaned up ([issues/128790#issuecomment-2590034178](https://github.com/kubernetes/kubernetes/issues/128790#issuecomment-2590034178)), I reverted all the eviction code for the terminated pods and only added a test that validates the cleanup of the terminated containers, checking that their resources are cleaned up correctly and allow disk pressure to be removed without affecting the running pods.

AnishShah commented:
/test pr-cos-cgroupv2-containerd-node-e2e-eviction
/test pr-crio-cgroupv2-node-e2e-eviction

AnishShah commented on test/e2e_node/eviction_test.go:
```diff
@@ -128,7 +129,47 @@ var _ = SIGDescribe("ImageGCNoEviction", framework.WithSlow(), framework.WithSer
+			// Setting the volume as nil, will make the file to be written to the
+			// container writable layer, which defaults to using the node's root fs.
+			{
+				evictionPriority: 0,
```
This is expected because we clean up the resource but not the pod?

KevinTMtz replied:
Yes, the pod is not evicted, its containers are just cleaned.

AnishShah commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

AnishShah commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

AnishShah commented:
lgtm overall. Can you fix the lint errors?

KevinTMtz commented:
> lgtm overall. Can you fix the lint errors?

Done @AnishShah

AnishShah commented:
/lgtm

yujuhong commented:
cc @tallclair or @SergeyKanzhelev for test review/approval

KevinTMtz commented:
/cc @tallclair

tallclair commented:
Need to add to the PR description
````
```release-note
NONE
```
````

tallclair commented:
/assign

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -576,10 +617,36 @@ func runEvictionTest(f *framework.Framework, pressureTimeout time.Duration, expe
+			// Wait for the expected succeeded pods to get to that state
+			for _, podToSucceed := range podsToSucceed {
+				ginkgo.By(fmt.Sprintf("Waiting for the %s pod to complete", podToSucceed.Name))
+				gomega.Eventually(ctx, func() error {
```
Use `framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, podToSucceed.Name, podToSucceed.Namespace))` instead

KevinTMtz replied:
Done, thank you.

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -736,8 +803,10 @@ func verifyEvictionOrdering(ctx context.Context, f *framework.Framework, testSpe
 		gomega.Expect(priorityPod).NotTo(gomega.BeNil())
-		gomega.Expect(priorityPod.Status.Phase).ToNot(gomega.Equal(v1.PodSucceeded),
-			fmt.Sprintf("pod: %s succeeded unexpectedly", priorityPod.Name))
+		if !isPodExpectedToSucceed(&priorityPod) {
+			gomega.Expect(priorityPod.Status.Phase).ToNot(gomega.Equal(v1.PodSucceeded),
+				fmt.Sprintf("pod: %s succeeded unexpectedly", priorityPod.Name))
+		}
```
Is it OK if the pod is terminated with a failure in this case? Or should it be running?

KevinTMtz replied:
Yes, the expected behavior would be to finish with a failure (for pods expected to be evicted) or to continue running (for pods that are expected to run forever). No other pod besides the red ones with `sleepAfterExecuting` equal to `true`, should succeed.

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -1033,32 +1102,32 @@ const (
 
 // podWithCommand returns a pod with the provided volumeSource and resourceRequirements.
-func podWithCommand(volumeSource *v1.VolumeSource, resources v1.ResourceRequirements, iterations int, name, command string) *v1.Pod {
+func podWithCommand(volumeSource *v1.VolumeSource, resources v1.ResourceRequirements, iterations int, name, command string, putToSleep bool) *v1.Pod {
```
nit: I'd prefer to call this parameter `runForever` (or `sleepForever`) and invert the logic. I think it's more clear what it's doing in that case. Alternatively, I'd also accept something like `sleepAfterExecuting`.

KevinTMtz replied:
Done, used `sleepAfterExecuting`, since it is more clear about what it does.

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -128,7 +129,47 @@ var _ = SIGDescribe("ImageGCNoEviction", framework.WithSlow(), framework.WithSer
+var _ = SIGDescribe("ImageGCTerminatedPodsContainersCleanup", framework.WithSlow(), framework.WithSerial(), framework.WithDisruptive(), feature.Eviction, func() {
+	f := framework.NewDefaultFramework("image-gc-terminated-pods-containers-cleanup-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+	pressureTimeout := 10 * time.Minute
```
looks like all the eviction tests are using this same timeout? Better to make it a shared constant.

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -128,7 +129,47 @@ var _ = SIGDescribe("ImageGCNoEviction", framework.WithSlow(), framework.WithSer
+	f := framework.NewDefaultFramework("image-gc-terminated-pods-containers-cleanup-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+	pressureTimeout := 10 * time.Minute
+	expectedNodeCondition := v1.NodeDiskPressure
+	expectedStarvedResource := resourceInodes
+	inodesConsumed := uint64(100000)
```
nit: const

KevinTMtz replied:
Done, made the change for all tests

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -128,7 +129,47 @@ var _ = SIGDescribe("ImageGCNoEviction", framework.WithSlow(), framework.WithSer
+
+	ginkgo.Context(fmt.Sprintf(testContextFmt, expectedNodeCondition), func() {
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			// Set the eviction threshold to inodesFree - inodesConsumed, so that using inodesConsumed causes an eviction.
```
I see that this is how other eviction tests are written, but this seems like it would be flaky without some extra padding. If some other process on the node freed an inode between now and the test execution, wouldn't the test fail?

KevinTMtz replied:
For this and the previous test, both have 10,000 inodes of padding, the tests will actually consume 110,000, but the threshold will be at 100,000. The constant's name `inodesConsumed` causes confusion, I actually represents the inodes that need to be consumed to surpass the threshold, I will rename it to `inodesToThreshold` to make it clear

KevinTMtz commented on test/e2e_node/eviction_test.go:
```diff
@@ -576,10 +617,36 @@ func runEvictionTest(f *framework.Framework, pressureTimeout time.Duration, expe
 			pods := []*v1.Pod{}
+			podsToSucceed := []*v1.Pod{}
 			for _, spec := range testSpecs {
+				if isPodExpectedToSucceed(spec.pod) {
```
Agree, done, thank you.

KevinTMtz commented on test/e2e_node/eviction_test.go:
```diff
@@ -576,10 +617,36 @@ func runEvictionTest(f *framework.Framework, pressureTimeout time.Duration, expe
 			}
-			e2epod.NewPodClient(f).CreateBatch(ctx, pods)
+			e2ePodClient := e2epod.NewPodClient(f)
+			e2ePodClient.CreateBatch(ctx, podsToSucceed)
```
Done, thank you. Added a comment of why `CreateBatch` was not used for future context.

KevinTMtz committed:
ImageGCTerminatedPodsContainersCleanup e2e node test

Updated ImageGCTerminatedPodsEviction to ImageGCTerminatedPodsContainersCleanup to test that
terminated containers are being cleaned up, instead of testing if terminated pods were being evicted

KevinTMtz commented:
/retest

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -591,10 +633,30 @@ func runEvictionTest(f *framework.Framework, pressureTimeout time.Duration, expe
+			// CreateBatch was not used to prevent a race condition because of
+			// underlying call to CreateSync, which may fail if the wait function
+			// misses to see the pod readiness before it terminates.
+			for _, pod := range podsToSucceed {
+				newPod := e2ePodClient.Create(ctx, pod)
+
+				ginkgo.By(fmt.Sprintf("Waiting for the %s pod to complete", newPod.Name))
+				framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, newPod.Name, newPod.Namespace))
+			}
```
Does this need to be done in serial? It would be faster to create all the pods before waiting:
```suggestion
			for _, pod := range podsToSucceed {
				newPod := e2ePodClient.Create(ctx, pod)
			}

			ginkgo.By(fmt.Sprintf("Waiting for the expected pods to complete: %v", podsToSucceed))
			for _, pod := range podsToSucceed {
				framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, newPod.Name, newPod.Namespace))
			}
```

KevinTMtz replied:
Done, thank you. 

I had to reassign the underlying value at the pod pointer `*pod = *e2ePodClient.Create(ctx, pod)`, so `e2epod.WaitForPodSuccessInNamespace` could find the pod. If I did not, this error would be produced:
```
[BeforeEach] k8s.io/kubernetes/test/e2e_node/eviction_test.go:628
  [It] k8s.io/kubernetes/test/e2e_node/eviction_test.go:671

  [FAILED] Timed out after 300.002s.
  The function passed to Eventually returned the following error:
      <framework.transientError>: 
      the server could not find the requested resource
      {
          error: <*errors.StatusError | 0xc0006e5180>{
              ErrStatus: {
                  TypeMeta: {Kind: "", APIVersion: ""},
                  ListMeta: {
                      SelfLink: "",
                      ResourceVersion: "",
                      Continue: "",
                      RemainingItemCount: nil,
                  },
                  Status: "Failure",
                  Message: "the server could not find the requested resource",
                  Reason: "NotFound",
                  Details: {Name: "", Group: "", Kind: "", UID: "", Causes: nil, RetryAfterSeconds: 0},
                  Code: 404,
              },
          },
      }
  In [BeforeEach] at: k8s.io/kubernetes/test/e2e_node/eviction_test.go:664 @ 04/01/25 00:27:12.977
```

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -591,10 +633,32 @@ func runEvictionTest(f *framework.Framework, pressureTimeout time.Duration, expe
+			// CreateBatch was not used to prevent a race condition because of
+			// underlying call to CreateSync, which may fail if the wait function
+			// misses to see the pod readiness before it terminates.
+			for _, pod := range podsToSucceed {
+				*pod = *e2ePodClient.Create(ctx, pod)
+			}
+
+			ginkgo.By(fmt.Sprintf("Waiting for the expected pods to complete: %v", podsToSucceed))
+			for _, pod := range podsToSucceed {
+				framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace))
+			}
```
(optional) It might be cleaner to just launch a bunch of goroutines:

```suggestion
			wg := sync.WaitGroup{}
			wg.Add(len(podsToSucceed))
			for _, pod := range podsToSucceed {
				go func(pod *v1.Pod) {
					defer wg.Done()
					defer ginkgo.GinkgoRecover()
					pod = e2ePodClient.Create(ctx, pod)
					framework.ExpectNoError(e2epod.WaitForPodSuccessInNamespace(ctx, f.ClientSet, pod.Name, pod.Namespace))
				}(pod)
			}
			wg.Wait()
```

It also means that each pod only waits for the maximum timeout. Otherwise, the last pod in the list gets to wait for the maximum timeout plus the aggregate wait time of all pods before it.

tallclair commented on test/e2e_node/eviction_test.go:
```diff
@@ -1083,7 +1149,7 @@ func podWithCommand(volumeSource *v1.VolumeSource, resources v1.ResourceRequirem
 						"sh",
 						"-c",
-						fmt.Sprintf("i=0; while [ $i -lt %d ]; do %s i=$(($i+1)); done; while true; do sleep 5; done", iterations, command),
+						fmt.Sprintf("i=0; while [ $i -lt %d ]; do %s i=$(($i+1)); done; while %v; do sleep 5; done", iterations, command, sleepAfterExecuting),
```
I recently learned that not trapping `TERM` here causes pause deletion to take a lot longer. We have a utility for that:

```go
cmd := fmt.Sprintf("i=0; while [ $i -lt %d ]; do %s i=$(($i+1)); done", iterations, command)
if sleepAfterExecuting {
	cmd += "; " + e2epod.InfiniteSleepCommand
}
```

KevinTMtz replied:
Done, thank you.

KevinTMtz commented:
/retest

tallclair commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

tallclair commented:
/lgtm

But I'm worried about flakes. Giving the eviction tests a retry.

tallclair commented:
/hold

@KevinTMtz can you investigate why those test suites are failing?

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz committed:
Remove sleepAfterExecuting param from diskConsumingPod

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz committed:
Revert trapping TERM for podWithCommand

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz commented:
https://github.com/kubernetes/kubernetes/pull/129240#issuecomment-2842658419
> @KevinTMtz can you investigate why those test suites are failing?

The test failures `ImageGCNoEviction` & `ImageGCTerminatedPodsContainersCleanup` are due to a race condition between the eviction cycles and and the cleaning of the resources, refer to:
https://github.com/kubernetes/kubernetes/blob/da24cfe98b3a4528770a0a5d9839b2172c266691/pkg/kubelet/eviction/helpers.go#L1195

The tests behave inconsistently, sometimes the resource reclaim finishes in the same eviction cycle, however some other times the resource reclaim just does not finish quickly enough to free resources and prevent a pod eviction in the same cycle, refer to:
https://github.com/kubernetes/kubernetes/blob/da24cfe98b3a4528770a0a5d9839b2172c266691/pkg/kubelet/eviction/eviction_manager.go#L382

This issue is tracking the current [failures](https://testgrid.k8s.io/sig-node-containerd#node-kubelet-containerd-eviction) of `ImageGCNoEviction`: https://github.com/kubernetes/kubernetes/issues/130663

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz commented:
/retest

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

KevinTMtz commented:
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-eviction
/test pull-crio-cgroupv2-node-e2e-eviction

SergeyKanzhelev commented on pkg/kubelet/eviction/eviction_manager_test.go:
```diff
@@ -1890,11 +1890,12 @@ func TestNodeReclaimFuncs(t *testing.T) {
 			fakeClock.Step(1 * time.Minute)
 
 			setDiskStatsBasedOnFs := func(whichFs string, diskPressure string, diskStat diskStats) diskStats {
-				if tc.inducePressureOnWhichFs == "nodefs" {
```
please do not include unrelated changes in the PR

KevinTMtz replied:
Done, thank you

KevinTMtz commented:
@SergeyKanzhelev @tallclair

> @KevinTMtz where are we on this?

It is ready for a final review, tests are passing successfully now (refer to [prow.k8s.io/pr-history/129240](https://prow.k8s.io/pr-history/?org=kubernetes&repo=kubernetes&pr=129240)).

I was waiting for the ImageGCNoEviction fix which was failing consistently (refer to https://github.com/kubernetes/kubernetes/pull/129240#issuecomment-2960741922), however that test is no longer failing, a different thing than the one I was waiting for may have fixed it (about 2 weeks ago).

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
